### DISCUSSION TITLE
Difference between `dask.array.image` and `dask_image`

### DISCUSSION THREAD

**Discussion by martinschorb** [dask/dask#7573] [Q&A]
Hi,

I discovered that there is an `imread` function in both `dask.array.image` and the separate `dask_image` package.
This duplication is a bit confusing. 
I assume `dask_image` provides more image-related functionality than the built-in function, but what exactly is the purpose of having two?

**Comment by jrbourbeau** [comment]
cc'ing @jakirkham @GenevieveBuckley

**Reply by GenevieveBuckley** [reply]
Hi @martinschorb, you're right: it is confusing to have this duplication.

Some background: Matt introduced `dask.array.image.imread()` to dask in 2015, and in 2017 @jakirkham started work on what would eventually become the `dask-image` library.

You can read John's explanation of the difference between the two here:

https://github.com/dask/dask-image/issues/84

>> I was wondering what the relationship between this package and the function dask.array.image.imread that's already part of dask.
>>
>> Especially as I detected that dask.array.image.imread doesn't actually make use of the remote data, so I couldn't give it a s3:// protocol.
>
>  jakirkham commented on 30 Oct 2018
>
> Sure. So this package provides a suite of image processing tools in the spirit of scipy.ndimage. Certainly loading image data into Dask is one piece of functionality it provides. Though most of the functionality is different types of image processing techniques that one may want to apply to the data once loaded into Dask Arrays (however one might choose to do that). It may be the case in the future that existing image utilities in Dask are relocated to or deprecated in favor of dask-image as was done with dask-ml, but we are not quite there yet.
>
> To be more specific, dask-image includes a large variety of filters (e.g. smoothing, denoising, edge detection, Fourier, etc.), morphological operations, and some operations on label images. There is certainly room for this package to grow in these areas based on the needs of the community. The functionality here is designed specifically to handle the fact that not all data needed for an operation may be in the same chunk. So will add overlaps with other chunks for filters or pull out relevant pieces of different chunks when working with label images. Hopefully that clarifies what dask-image is trying to solve and how it differs from functionality in Dask.
>
> Loading image data is generally a hard problem (even outside of Dask) due to the large variety of formats, image format extensions or specializations for specific fields, the requirements of different imaging modalities, file constraints (size, dimensionality, ordering, etc.), compression, access patterns, encryption, etc. As a consequence there are more than a few libraries that can be used to load image data with various tradeoffs that range between how closely the loaded data should match the format to smoothing out differences between many different formats by loading array data generally.
> 
> For imread specifically, dask-image has made some choices that are different from the imread function in Dask. These are done to improve graph construction performance. In either case, the actual data loading step is handed off to an external library. In Dask's case, this is scikit-image or a user defined function. In dask-image, this is PIMS, which then can use any of a number of different things including (but not limited to) scikit-image depending on what is installed and available to it. Both can be reasonable choices.
>
> Would be happy to discuss the particular problem you are working with, but would need a bit more detail. Namely what image formats are involved, how the data is split up (if at all), whether authentication of some kind is needed, etc. There could be short term solutions using things like dask.delayed, S3fs, s3fs-fuse, etc., which would give you a way to load data that you need to analyze now. Longer term solutions would be integrating these things in a more appetizing way for users like yourself.

https://github.com/dask/dask-image/issues/134#issuecomment-593195715

>>    I heard second-hand that PIMS was chosen because it defers computation more than other readers and is thus suitable for dask, but I can't speak to that authoritatively.
> 
>  jakirkham commented on 2 Mar 2020
>
> That's right :smile: (authoritatively confirmed wink)

**Reply by GenevieveBuckley** [reply]
My personal opinion is that it would be good if we could move towards a situation that's more like the zen of python suggestion: "There should be one-- and preferably only one --obvious way to do it."

However that is a little tricky to do, since it's not especially clear what is the one "best" way of doing things - and how much what is "best" changes under different circumstances. There have been some attempts to compare the performance of different `imread` methods. It's likely more benchmarking would be needed to resolve this questions.

* https://github.com/dask/dask-image/issues/121
* https://github.com/dask/dask-image/issues/181
